diff --git a/webrtc/modules/video_coding/codecs/h264/h264_encoder_impl.cc b/webrtc/modules/video_coding/codecs/h264/h264_encoder_impl.cc
index d361b8a4f..978e90181 100644
--- a/webrtc/modules/video_coding/codecs/h264/h264_encoder_impl.cc
+++ b/webrtc/modules/video_coding/codecs/h264/h264_encoder_impl.cc
@@ -376,7 +376,10 @@ NVENCSTATUS H264EncoderImpl::Deinitialize()
 	ReleaseIOBuffers();
 	nvStatus = m_pNvHWEncoder->NvEncDestroyEncoder();
 
-	__cu(cuCtxDestroy(m_cuContext));
+	if (!m_d3dDevice)
+	{
+		__cu(cuCtxDestroy(m_cuContext));
+	}
 	return nvStatus;
 }
 
